sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -487,42 +487,81 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
     case u: UnresolvedAttribute =>
-      resolveUnresolvedAttributeByPlanId(
-        u, q, idToPlan: mutable.HashMap[Long, LogicalPlan]
-      ).getOrElse(u)
+      resolveUnresolvedAttributeByPlanId(u, q).getOrElse(u)
     case _ if e.containsPattern(UNRESOLVED_ATTRIBUTE) =>
-      e.mapChildren(c => tryResolveColumnByPlanId(c, q, idToPlan))
+      e.mapChildren(c => tryResolveColumnByPlanId(c, q))
     case _ => e
   }
 
   private def resolveUnresolvedAttributeByPlanId(
       u: UnresolvedAttribute,
-      q: LogicalPlan,
-      idToPlan: mutable.HashMap[Long, LogicalPlan]): Option[NamedExpression] = {
+      q: LogicalPlan): Option[NamedExpression] = {
     val planIdOpt = u.getTagValue(LogicalPlan.PLAN_ID_TAG)
     if (planIdOpt.isEmpty) return None
     val planId = planIdOpt.get
     logDebug(s"Extract plan_id $planId from $u")
 
-    val plan = idToPlan.getOrElseUpdate(planId, {
-      findPlanById(u, planId, q).getOrElse {
-        // For example:
-        //  df1 = spark.createDataFrame([Row(a = 1, b = 2, c = 3)]])
-        //  df2 = spark.createDataFrame([Row(a = 1, b = 2)]])
-        //  df1.select(df2.a)   <-   illegal reference df2.a
-        throw new AnalysisException(
-          errorClass = "_LEGACY_ERROR_TEMP_3051",
-          messageParameters = Map(
-            "u" -> u.toString,
-            "planId" -> planId.toString,
-            "q" -> q.toString))
+    val isMetadataAccess = u.getTagValue(LogicalPlan.IS_METADATA_COL).isDefined
+
+    var result = Option.empty[NamedExpression]
+    var numMatched = 0
+    resolveUnresolvedAttributeByPlanId(u, planId, isMetadataAccess, q).foreach {
+      case Some(resolved) =>
+        numMatched += 1
+        if (result.isEmpty) {
+          result = Some(resolved)
+        } else {
+          throw new AnalysisException(
+            errorClass = "AMBIGUOUS_COLUMN_REFERENCE",
+            messageParameters = Map("name" -> toSQLId(u.nameParts)),
+            origin = u.origin
+          )
+        }
+      case _ => numMatched += 1
+    }
+    if (numMatched == 0) {
+      // For example:
+      //  df1 = spark.createDataFrame([Row(a = 1, b = 2, c = 3)]])
+      //  df2 = spark.createDataFrame([Row(a = 1, b = 2)]])
+      //  df1.select(df2.a)   <-   illegal reference df2.a
+      throw new AnalysisException(
+        errorClass = "_LEGACY_ERROR_TEMP_3051",
+        messageParameters = Map(
+          "u" -> u.toString,
+          "planId" -> planId.toString,
+          "q" -> q.toString))
+    }
+    result
+  }
+
+  private def resolveUnresolvedAttributeByPlanId(
+      u: UnresolvedAttribute,
+      id: Long,
+      isMetadataAccess: Boolean,
+      p: LogicalPlan): Iterator[Option[NamedExpression]] = {
+    if (p.getTagValue(LogicalPlan.PLAN_ID_TAG).contains(id)) {
+      Iterator.single(
+        resolveUnresolvedAttributeByPlan(u, p, isMetadataAccess))
+    } else {
+      p.children.iterator.flatMap { child =>
+        val outputSet = if (isMetadataAccess) {
```

Comment:
I think `resolveUnresolvedAttributeByPlanId` should always return a valid column. So the code can be further simplified:
```
def resolveUnresolvedAttributeByPlanId(...): Option[NamedExpression] = {
  ...
  val candidates = p.children.flatMap(resolveUnresolvedAttributeByPlanId(...))
    .filter(p.outputSet.contains)
  if (candidates.length > 1) fail ...
  candidates.headOption
}
```

Reply:
This is more efficient as plans like `Project` can prune the resolved column earlier. The outer `resolveUnresolvedAttributeByPlanId` should fail if there is no column matched.